task/sensor/src/main.rs

```diff
@@ -49,11 +66,15 @@ impl idl::InOrderSensorImpl for ServerImpl {
         _: &RecvMessage,
         id: SensorId,
         value: f32,
+        timestamp: u64,
     ) -> Result<(), RequestError<SensorError>> {
         let index = id.0;
 
         if index < NUM_SENSORS {
-            self.data[index] = Reading::Value(value);
+            self.has_data[index] = true;
+            self.data_value[index] = value;
+            self.data_time[index] = timestamp;
+            self.nodata[index] = None;
```

Comment:
If you timestamp the nodata reading, I would suggest removing this line so we can see the last error even if there's been a success since then.

Reply:
If we do that, do we need separate functions for "get last valid reading (if any)", "get last error (if any)", and "get last reading (if any, of either kind)"?

Reply:
I decided to change this in two ways:

- Adding timestamps to the error readings (and not clearing the error readings when writing data)
- changing `has_data` to an `enum` representing whether the `data` or `err` arrays were last written

`get()` and `get_reading()` still have the semantics of "get last reading (of either kind)", chosen by this enum.

I'd vote to add functions with the other semantics once they're needed (in MGS work or elsewhere).